[code-completion] Add type context for single-expression closures

benlangmuir opened the pull request:
When completing in the only expression of closure, use the return type of the closure as the type context for the code-completion. However, since code-completion may be on an incomplete input, we only use the return type to improve the quality of the result, not to mark it invalid, since (a) we may add another statement afterwards, or (b) if the context type is Void it doesn't need to match the value.

benlangmuir committed:
[code-completion] Add type context for single-expression closures

When completing in the only expression of closure, use the return type
of the closure as the type context for the code-completion. However,
since code-completion may be on an incomplete input, we only use the
return type to improve the quality of the result, not to mark it
invalid, since (a) we may add another statement afterwards, or (b) if
the context type is Void it doesn't need to match the value.

benlangmuir commented:
@swift-ci please test

benlangmuir commented:
@rintaro this is a bit different than we discussed.  Instead of handling Void specially, I keep track of whether the type context is from a single-expression body and in that case do not mark anything as type-relation "invalid".

benlangmuir committed:
Fix lifetime of type context in result builder

There usually won't be a lot of expected types, so just use a
SmallVector and copy it instead of trying to get tricky with the
lifetime.

benlangmuir commented:
@swift-ci please test

benlangmuir committed:
Fix assertion failure in validation test

Seen in validation-test/IDE/crashers_fixed/019-swift-vardecl-emitlettovarnoteifsimple.swift

rintaro commented on lib/IDE/ExprContextAnalysis.cpp:
```diff
@@ -705,14 +713,28 @@ class ExprContextAnalyzer {
+    case 1:
+      return body->getElements()[0].is<Expr *>();
+    default:
+      return false;
```
Do we already have test cases for this path? like:
```swift
var floatVal: Float = 2
let _: () -> Float = {
  #^COMPLETE^# // Should be normal completion.
  return 42
}
let _: () -> SomeEnum = {
  .#^COMPLETE^# // Should not provide anything.
  return 42
}
```
If not, please add them.

benlangmuir replied:
Good catch, this code for handling `getNumElements() == 0` is from before I changed the parser to add the `CodeCompletionExpr` in `completeStmtOrExpr`.  I'll remove the 0 case and add the additional test.

benlangmuir commented:
@swift-ci please test

rintaro reviewed (commented):
Implementation makes sense to me :)

benlangmuir committed:
Add tests for non-single-expression closures and remove dead code

Ensure that we get the correct behaviour when we are not in the
single-expression case, and remove code that handled 0-element bodies,
which is no longer needed after the parser change.

benlangmuir commented:
@swift-ci please test

benlangmuir commented:
@swift-ci please test macOS
